contracts/snowcones/gauge-proxy-v3.sol

```diff
@@ -0,0 +1,806 @@
+    }
+
+    function getRewardForDuration() external view returns (uint256) {
+        return rewardRate.mul(DURATION);
+    }
+    
+    function depositAll() external {
+        _deposit(TOKEN.balanceOf(msg.sender), msg.sender);
+    }
+    
+    function deposit(uint256 amount) external {
+        _deposit(amount, msg.sender);
+    }
+    
+    function depositFor(uint256 amount, address account) external {
+        _deposit(amount, account);
+    }
+    
+    function _deposit(uint amount, address account) internal nonReentrant updateReward(account) {
+        require(amount > 0, "Cannot stake 0");
+        _totalSupply = _totalSupply.add(amount);
+        _balances[account] = _balances[account].add(amount);
+        emit Staked(account, amount);
+        TOKEN.safeTransferFrom(account, address(this), amount);
+    }
+    
+    function withdrawAll() external {
+        _withdraw(_balances[msg.sender]);
+    }
+
+    function withdraw(uint256 amount) external {
+        _withdraw(amount);
+    }
+    
+    function _withdraw(uint amount) internal nonReentrant updateReward(msg.sender) {
+        require(amount > 0, "Cannot withdraw 0");
+        _totalSupply = _totalSupply.sub(amount);
+        _balances[msg.sender] = _balances[msg.sender].sub(amount);
+        TOKEN.safeTransfer(msg.sender, amount);
+        emit Withdrawn(msg.sender, amount);
+    }
+
+    function getReward() public nonReentrant updateReward(msg.sender) {
+        uint256 reward = rewards[msg.sender];
+        if (reward > 0) {
+            rewards[msg.sender] = 0;
+            SNOWBALL.safeTransfer(msg.sender, reward);
+            emit RewardPaid(msg.sender, reward);
+        }
+    }
+
+    function exit() external {
+       _withdraw(_balances[msg.sender]);
+        getReward();
+    }
+    
+    function notifyRewardAmount(uint256 reward) external onlyDistribution updateReward(address(0)) {
+        SNOWBALL.safeTransferFrom(DISTRIBUTION, address(this), reward);
+        if (block.timestamp >= periodFinish) {
+            rewardRate = reward.div(DURATION);
+        } else {
+            uint256 remaining = periodFinish.sub(block.timestamp);
+            uint256 leftover = remaining.mul(rewardRate);
+            rewardRate = reward.add(leftover).div(DURATION);
+        }
+
+        // Ensure the provided reward amount is not more than the balance in the contract.
+        // This keeps the reward rate in the right range, preventing overflows due to
+        // very high values of rewardRate in the earned and rewardsPerToken functions;
+        // Reward + leftover must be less than 2^256 / 10^18 to avoid overflow.
+        uint balance = SNOWBALL.balanceOf(address(this));
+        require(rewardRate <= balance.div(DURATION), "Provided reward too high");
+
+        lastUpdateTime = block.timestamp;
+        periodFinish = block.timestamp.add(DURATION);
+        emit RewardAdded(reward);
+    }
+
+    modifier updateReward(address account) {
```

Comment:
For consistency, modifier could be grouped together and declared before any function.